Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -371,9 +431,231 @@ def checkFileLength (lines : Array String) (existingLimit : Option ℕ) : Option
+'⊲', 'ϑ', 'ϐ', '⊳', '⊻', 'ě', 'Ě', 'ď', '⋮', 'Ď', 'Č',
+'č', '₭', 'ϟ', 'Į', 'į', 'K', '⚠', 'ϧ', '≀', '℘', 'Ϯ',
+'Ϝ', 'Ð', 'Η', '≎', '𝔻', '𝔼', '𝔾', '𝕁', '𝕀', '𝕃', '𝕄',
+'𝕆', '𝕋', '𝕊', '𝕍', '𝕌', '𝕏', '𝕎', '𝕐', '𝕓', '𝕒', '𝕕',
+'𝕔', '𝕗', '𝕖', '𝕙', '𝕘', '𝕛', '𝕚', '𝕜', '𝕟', '𝕞', '𝕡',
+'𝕠', '𝕣', '𝕢', '𝕥', '𝕤', '𝕧', '𝕦', '𝕩', '𝕨', '𝕪', '𝕫',
+'⨯', '⨿', 'Ϳ'
+]
+
+/-- Other characters already in Mathlib as of Aug. 28, 2024 -/
+def othersInMathlib := #[
+'🔍', '🐙', '💡', '▼', 'c', 'ō', '🏁', '⏳', '⏩', '❓',
+'🆕', 'š', 'ř', '✅', '❌', '⚬', '│', '├', '┌', 'ő',
+'⟍', '̂', 'ᘁ', 'ń', 'ć', '⟋', 'ỳ', 'ầ', '⥥',
+'ł', '◿', '◹', '－', '＼', '◥', '／', '◢', 'Ž', 'ă',
+'И', 'в', 'а', 'н', 'о', 'и', 'ч', 'Š', 'ᴜ', 'ᵧ', '´',
+'ᴄ', 'ꜰ', 'ß', 'ᴢ', 'ᴏ', 'ᴀ', 'ꜱ', 'ɴ', 'ꞯ', 'ʟ',
+'ʜ', 'ᵟ', 'ʙ', 'ᵪ', 'ᵩ', 'ᵦ', 'ᴊ', 'ᴛ', 'ᴡ', 'ᴠ',
+'ɪ', '̀', 'ᴇ', 'ᴍ', 'ʀ', 'ᴅ', 'ɢ', 'ʏ', 'ᴘ', 'ĝ', 'ᵨ',
+'ᴋ', 'ś', '꙳', '𝓡', '𝕝', '𝖣', '⨳', '🎉', '🟡']
+
+/- TODO there are more symbols we could use that aren't in this list yet. E.g, see
+ https://en.wikipedia.org/wiki/Mathematical_operators_and_symbols_in_Unicode
+-/
+
+/-- Unicode symbols in Mathilb that should always be followed by the emoji-variant selector -/
+def mathlibEmojiSymbols := #[
+'✅', '❌', '💥', '🟡', '💡', '🐙', '🔍', '🎉', '⏳', '🏁'
+]
+
+/-- Unicode symbols in mathilb that should always be followed by the text-variant selector -/
+def mathlibTextSymbols : Array Char := #[]
+
+/--
+Hash-set of all unicode characters allowed by the unicodeLinter.
+-/
+def unicodeWhitelist : Lean.HashSet Char := Lean.HashSet.ofArray (
+    printableASCII ++ withVSCodeAbbrev ++ othersInMathlib ++
+    #[UnicodeVariant.emoji, UnicodeVariant.text]
+)
+
+/-- Checks if a character is unrecommended, according to the unicodeLinter (`unwantedUnicode`)-/
+def isBadChar (c : Char) : Bool := !unicodeWhitelist.contains c
+
+end unicodeLinter
+
+open unicodeLinter in
+
+/-- Lint a collection of input strings if one of them contains unwanted unicode. -/
+def unicodeLinter : TextbasedLinter := fun lines ↦ Id.run do
+  let mut errors : Array (StyleError × ℕ) := Array.mkEmpty 0
+  let mut lineNumber := 1
+  for line in lines do
+    -- Ensure no bad unicode characters are used
+    let badChars := line.toList.filter isBadChar
+    if badChars.length > 0 then
+      let newErrors : Array (StyleError × ℕ) :=
+        badChars.toArray.map (StyleError.unwantedUnicode ·, lineNumber)
+      errors := errors.append newErrors
+
+    for (selector, symbols) in #[
+        (UnicodeVariant.emoji, mathlibEmojiSymbols),
+        (UnicodeVariant.text, mathlibTextSymbols) ] do
+      -- Ensure specified emojis/text-symbols have the correct variant-selector
+      let charPos := line.findAll (symbols.contains ·)
+      for pos in charPos do
+        let nextC := line.get₂ (line.next pos)
+        let previousC := line.get₂ (line.prev pos) -- may be the char. itself if `pos == 0`
+        unless (previousC == '\'' ∧ nextC == '\'') ∨ nextC == selector do
```

Comment:
What is the logic here? What do the ' characters signify?

Reply:
outdated now